### PR TITLE
Migrations of task configs using an int scheme

### PR DESCRIPTION
**Pull request by gdraynz**

### CODE REVIEW

File: nyuki/workflow/tasks/trigger_workflow.py [hunk 1]
```diff
@@ -16,6 +16,19 @@
 log = logging.getLogger(__name__)
 
 
+def _migrate_1(task):
+    config = task['config']
+    task['config'] = {
+        'template': {
+            'service': 'twilio' if 'twilio' in config['nyuki_api'] else 'pipeline',
```
Comment (lhavel) [inline nyuki/workflow/tasks/trigger_workflow.py]: Why do you set a `service` key for each task? There's already a `service` key at the workflow template level...

File: nyuki/workflow/tasks/trigger_workflow.py [hunk 2]
```diff
@@ -16,6 +16,19 @@
 log = logging.getLogger(__name__)
 
 
+def _migrate_1(task):
```
Comment (lhavel) [inline nyuki/workflow/tasks/trigger_workflow.py]: I don't feel comfortable with migrations described in referenced in a class that implements the business logic.
We'd rather design a clean db layer that deals with that stuff.
This is the major caveat I can see in your PR.

File: nyuki/workflow/tasks/trigger_workflow.py [hunk 3]
```diff
@@ -16,6 +16,19 @@
...
+    config = task['config']
+    task['config'] = {
+        'template': {
+            'service': 'twilio' if 'twilio' in config['nyuki_api'] else 'pipeline',
+            'id': config['template'],
+            'draft': config.get('draft', False),
```
Comment (lhavel) [inline nyuki/workflow/tasks/trigger_workflow.py]: Same question with `draft`

File: nyuki/workflow/tasks/trigger_workflow.py (hunk 1 above)
Comment (gdraynz) [inline nyuki/workflow/tasks/trigger_workflow.py]: It points to the template to be triggered by the `trigger_workflow` task.

File: nyuki/workflow/api/templates.py [hunk 4]
```diff
@@ -73,19 +102,27 @@ def __init__(self, templates_collection, metadata_collection):
                 template.update(metadatas[template['id']])
 
         if latest is False and draft is False:
+            if full is True:
+                for template in templates:
+                    await self._migrate(template)
             return templates
 
         # Retrieve the latest versions + drafts
         lasts = {}
```
Comment (gdraynz) [inline nyuki/workflow/api/templates.py]: The frontend does not use it, I sometimes do when in development.

File: nyuki/workflow/api/templates.py [hunk 5]
```diff
@@ -41,6 +43,33 @@ def __init__(self, templates_collection, metadata_collection):
...
+                # Do one step of migration.
+                do_migration(task)
+                task['scheme'] = scheme
+                must_save = True
+                log.info(
+                    'Migrated task %s of template %s from %s to %s',
+                    task['id'], template['id'], old_scheme, scheme,
+                )
+
+        if must_save:
```
Comment (gdraynz) [inline nyuki/workflow/api/templates.py]: This is to ensure we only replace the database document if at least one task was migrated.

File: nyuki/workflow/api/templates.py (hunk 4 above)
Comment (lhavel) [inline nyuki/workflow/api/templates.py]: I'm ok to improve Mongo requests in a further issue.
But we should not leave unused options in the code. If you need full details of a workflow template the `get` method is the one you should use.

File: nyuki/workflow/api/templates.py (hunk 5 above)
Comment (lhavel) [inline nyuki/workflow/api/templates.py]: yep, and the `if migrations[scheme:]` condition evaluates to true only if there's at least one migration to apply and no exception occurred during the migration process. This is what you want isn't it?

File: nyuki/workflow/tasks/trigger_workflow.py (hunk 2 above)
Comment (gdraynz) [inline nyuki/workflow/tasks/trigger_workflow.py]: This is from my point of view the safest and simplest spot to never miss a migration when one of our tasks has to change: everything in the corresponding task's module.
Please note that the actual "database model" of each task really is the attached JsonSchema.
If I agree that our database model must be better controlled, I find this specific task migration system ok.

Comment (lhavel) [inline nyuki/workflow/tasks/trigger_workflow.py]: The task holder class is meant to execute the business logic.
It is neither meant to define the data served by the HTTP API and nor the data stored into the db backend.
I wish we get back to more future-proof design patterns.
The JSON schema may no longer be part of the task holder class in a near future...
I understand your _"everything in one place is simpler, safer"_ but that mixes distinct concerns

File: nyuki/workflow/api/templates.py [hunk 6]
```diff
@@ -41,6 +43,33 @@ def __init__(self, templates_collection, metadata_collection):
             [('id', DESCENDING), ('draft', DESCENDING)]
         ))
 
+    async def _migrate(self, template):
+        # Check and do migrations if needed
+        must_save = False
+        for task in template['tasks']:
+            old_scheme = task.get('scheme', 0)
```
Comment (lhavel) [inline nyuki/workflow/api/templates.py]: `old_scheme` => `cur_scheme`, the scheme value may not change

File: nyuki/workflow/api/templates.py (hunk 5 above)
Comment (lhavel) [inline nyuki/workflow/api/templates.py]: My bad. I didn't read the condition at the right indentation level.

File: nyuki/workflow/api/templates.py [hunk 7]
```diff
@@ -109,6 +146,37 @@ def __init__(self, templates_collection, metadata_collection):
                 for template in templates:
                     template.update(metadatas[0])
 
+        for template in templates:
```
Comment (lhavel) [inline nyuki/workflow/api/templates.py]: I didn't get it at once but each time you read templates, you try to migrate it!
Which means you need to walk through the list of templates and for each template you walk through the list of tasks to ensure there's nothing to migrate.
This is likely to add a significant overhead and may lead to performance issues.

Comment (lhavel) [inline nyuki/workflow/api/templates.py]: We may have to _flag_ each document as _migrated_ once it's done...

**Comment by lhavel** [comment]
We identified several problems in the approach proposed by @gdraynz.
This branch won't be merged.
New issues and branches shall be created to address migrations with a more future-proof approach.

**Pull request closed without merging**